src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,92 @@
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
-    private static final Injector injector = new Injector();
+    private static final Injector INJECTOR = new Injector();
+    private static final Map<Class<?>, Class<?>> INTERFACES_IMPLEMENTATIONS;
+    private static final Map<Class<?>, Object> INSTANCES = new HashMap<>();
+
+    static {
+        INTERFACES_IMPLEMENTATIONS = interfacesImplementationsConfig();
+    }
 
     public static Injector getInjector() {
-        return injector;
+        return INJECTOR;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] fields = clazz.getDeclaredFields();
+        for (Field field : fields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. Class: "
+                            + clazz.getName() + ", field: " + field.getName());
+                }
+            }
+        }
+        return clazzImplementationInstance == null ? createNewInstance(clazz)
+                : clazzImplementationInstance;
+    }
+
+    private static Object createNewInstance(Class<?> clazz) {
+        if (INSTANCES.containsKey(clazz)) {
+            return INSTANCES.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            INSTANCES.put(clazz, instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Can't create new instance of class - " + clazz.getName());
+        }
+    }
+
+    private static Class<?> findImplementation(Class<?> interfaceClazz) {
+        if (interfaceClazz.isInterface()) {
+            interfaceCheck(interfaceClazz);
+            interfaceClazz = INTERFACES_IMPLEMENTATIONS.get(interfaceClazz);
+        }
+        componentCheck(interfaceClazz);
+        return interfaceClazz;
+    }
+
+    private static Map<Class<?>, Class<?>> interfacesImplementationsConfig() {
+        Map<Class<?>, Class<?>> map = new HashMap<>();
+        map.put(FileReaderService.class, FileReaderServiceImpl.class);
+        map.put(ProductParser.class, ProductParserImpl.class);
+        map.put(ProductService.class, ProductServiceImpl.class);
+        return map;
+    }
```

Comment:
Common mistakes: "Make Interface Implementations map a class field. You can fill it in using Map.of()"